pkg/guardian/jail.go

```diff
@@ -0,0 +1,165 @@
+}
+
+// PrisonerStore can check if a client ip is a prisoner and also add prisoners
+type PrisonerStore interface {
+	IsPrisoner(remoteAddress string) bool
+	AddPrisoner(remoteAddress string, jail Jail)
+}
+
+// A JailProvider is a generic interface for determining which jail (if any) applies to a request
+type JailProvider interface {
+	GetJail(req Request) Jail
+}
+
+// A RouteJailStore can retrieve the Jail configuration for a given route
+type RouteJailStore interface {
+	GetJail(u url.URL) Jail
+}
+
+// A RouteJailProvider is a jail provider that provides a jail based off the route of a request
+type RouteJailProvider struct {
+	logger logrus.FieldLogger
+	store  RouteJailStore
+}
+
+// GetJail queries the store to determine which Jail, if any, applies to the request.
+func (jp *RouteJailProvider) GetJail(req Request) Jail {
+	reqUrl, err := url.Parse(req.Path)
+	if err != nil || reqUrl == nil {
+		return Jail{} // Limit disabled by default
+	}
+	jp.logger.Debugf("Getting jail for %v", reqUrl)
+	return jp.store.GetJail(*reqUrl)
+}
+
+// NewRouteRateLimitProvider returns an implementation of a LimitProvider intended to provide limits based off a
+// a request's path and a client's IP address.
+func NewRouteJailProvider(store RouteJailStore, logger logrus.FieldLogger) *RouteJailProvider {
+	return &RouteJailProvider{store: store, logger: logger}
+}
+
+// routeJailKeyFunc provides a key unique to a particular client IP and request path.
+func routeJailKeyFunc(req Request) string {
+	return "jailer:" + req.RemoteAddress + ":" + req.Path
+}
+
+type GenericJailer struct {
+	keyFunc      func(req Request) string
+	jailProvider JailProvider
+	store        PrisonerStore
+	counter      Counter
+	logger       logrus.FieldLogger
+	onJailsHandled []JailsHandledHook
+}
+
+type JailsHandledHook func(req Request, blocked bool, dur time.Duration, err error)
+
+func OnGenericJailerHandled(mr MetricReporter) JailsHandledHook {
+	return func(req Request, blocked bool, dur time.Duration, err error) {
+		opts := []MetricOptionSetter{}
+		if blocked {
+			// Only set additional tags for requests that were blocked. This way, we know the cardinality of the custom metrics.
+			opts = append(opts, WithRemoteAddress(req.RemoteAddress))
+		}
+		mr.HandledJail(req, blocked, err != nil, dur, opts...)
+	}
+}
+
+func (gj *GenericJailer) IsBanned(ctx context.Context, request Request) (banned bool, err error) {
+	if gj.keyFunc == nil || gj.jailProvider == nil || gj.counter == nil || gj.logger == nil || gj.store == nil {
+		gj.logger.Error("misconfigured generic jailer: missing required field")
+		return false, nil
+	}
+
+	start := time.Now().UTC()
+	defer func() {
+		for _, fn := range gj.onJailsHandled {
+			fn(request, banned, time.Since(start), err)
+		}
+	}()
+
+	banned = gj.store.IsPrisoner(request.RemoteAddress)
+	if banned {
+		return banned, nil
+	}
+
+	jail := gj.jailProvider.GetJail(request)
+	gj.logger.Debugf("fetched jail %v for request %v", jail, request.Path)
```

Comment:
we may not want these log calls in the hot path of every request

Reply:
Think we have traditionally had debug logs in the hot path of requests in Guardan (see examples in `ratelimit.go`). It is just not recommended to use a `debug` log level in production. 

If we do not want this behavior though, we can remove the logs from this PR and also audit the rest of the codebase to remove those other log calls as well.